### PR TITLE
redis: alternate healthcheck

### PR DESCRIPTION
**Pull request by danielhochman**
**Description**
Adds EXISTS healthcheck when key is specified in config. This allows the user to fail the healthcheck. See the data-plane changes linked below for more details.

**Risk Level**
Low

**Testing**
Unit, manual

**Docs and API Changes**
https://github.com/envoyproxy/data-plane-api/pull/499

### CODE REVIEW

File: source/common/upstream/health_checker_impl.h [hunk 1]
```diff
@@ -406,7 +406,7 @@ class RedisHealthCheckerImpl : public HealthCheckerImplBase {
   };
 
   struct HealthCheckRequest {
-    HealthCheckRequest();
+    HealthCheckRequest(const std::string& key);
```
Comment (jsedgwick) [inline source/common/upstream/health_checker_impl.h]: Using empty strings to toggle request type is a little wacky. Can you be more explicit and use subclasses (i.e. `PingHealthCheckRequest`)? Or keep the no-arg ctor for PING.

Comment (danielhochman) [reply source/common/upstream/health_checker_impl.h]: i was thinking of using `Optional` actually, thoughts on that vs no-arg ctor?

File: source/common/config/cds_json.cc [hunk 2]
```diff
@@ -53,7 +53,9 @@ void CdsJson::translateHealthCheck(const Json::Object& json_health_check,
     }
   } else {
     ASSERT(hc_type == "redis");
-    health_check.mutable_redis_health_check();
+    auto* redis_health_check = health_check.mutable_redis_health_check();
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/config/cds_json.cc]: @jsedgwick FWIW we do generally do this in the code base. I think the documentation is useful even though technically, yes, it doesn't do anything. So I would leave it. Also feel free to update STYLE.md in the auto section if desired. (I'm not a huge fan of auto for clarity reasons and doing things like this I think can somewhat help readability).

[the remark this answers is not in the document]
Comment (jsedgwick) [reply source/common/config/cds_json.cc]: That's fine. Though on `auto` in general, I'm in the Herb Sutter school (https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) but would refrain from recommending it in other people's code if you're of a different mind.

File: source/common/upstream/health_checker_impl.h (hunk 1 above)
Comment (jsedgwick) [reply source/common/upstream/health_checker_impl.h]: That solves the "empty string is magic" issue but still doesn't indicate what requests you're actually making. No-arg ctor or `HCR(Type, Optional<String>)`?
Where `HealthCheckRequest::Type` is `enum class Type { EXISTS, PING }`?

**Comment by danielhochman** [comment]
ready for re-review @htuch

**Pull request merged**